cartographer/mapping/internal/3d/pose_graph_3d.cc

```diff
@@ -509,40 +509,59 @@ void PoseGraph3D::HandleWorkQueue(
 }
 
 void PoseGraph3D::WaitForAllComputations() {
-  bool notification = false;
-  common::MutexLocker locker(&mutex_);
+  int num_trajectory_nodes;
+  {
+    common::MutexLocker locker(&mutex_);
+    num_trajectory_nodes = data_.num_trajectory_nodes;
+  }
+
   const int num_finished_nodes_at_start =
       constraint_builder_.GetNumFinishedNodes();
+
+  // First wait for the work queue to drain so that it's safe to schedule
+  // a WhenDone() callback.
+  {
+    common::MutexLocker locker(&work_queue_mutex_);
+    locker.Await([this]() REQUIRES(work_queue_mutex_) {
```

Comment:
This can take a long time. Can we switch this to a while loop with `AwaitWithTimeout()` similar to below? The functionality to output the progress can be deduplicated extracting it into a local function with a lambda.

Reply:
Done.